### PR TITLE
fix(spanner): support querying UUID against STRING columns by using untyped binding

### LINKED ISSUE #4120: spring-cloud-gcp-data-spanner error when querying by UUID
Java 21
Spring Boot 3.5.3
spring-cloud-gcp Version 7.0.0+

Prior to spring-cloud-gcp-data-spanner version 7.0.0, querying Spanner with a UUID successfully returned results. However, since upgrading the version to 7.0.0 or newer (7.3.0 newest version as of this post), an error is now present when trying to query with a UUID data type.

Error:
```
"INVALID_ARGUMENT: io.grpc.StatusRuntimeException: INVALID_ARGUMENT: No matching signature for operator = for argument types: STRING, UUID\n  Signature: T1 = T1\n    Unable to find common supertype for templated argument <T1>\n      Input types for <T1>: {STRING, UUID} [at 1:155]\n...name FROM my_table WHERE ( my_uuid=@tag0 )\n                                           ^ - Statement: 'SELECT my_id, my_uuid, name FROM my_table WHERE ( my_uuid=@tag0 )'",
```
This error is presented using SpannerRepository derived queries and native queries.

Sample code:
```
@Table(name = "my_table")
public class MyTableEntity {

    @PrimaryKey
    @Column(name = "my_id")
    private UUID myId;

    @Column(name = "my_uuid")
    private UUID myUuid;

    @Column(name = "name")
    private String name;

}
```
```
@Repository
public interface MyTableRepository extends SpannerRepository<MyTableEntity, UUID> {

    Optional<MyTableEntity> findByMyUuid(UUID myUuid);

}
```
Calling `myTableRepository.findByMyUuid(myUuid);` now throws the above error.

### PR DESCRIPTION
**Pull request by sakthivelmanii**
Binds java.util.UUID parameters as Untyped so that BE resolves the correct type which solves the issues with legacy schemas using STRING columns.

Fixes: https://github.com/GoogleCloudPlatform/spring-cloud-gcp/issues/4120

### CODE REVIEW

File: spring-cloud-gcp-data-spanner/src/main/java/com/google/cloud/spring/data/spanner/core/convert/ConverterAwareMappingSpannerEntityWriter.java [hunk 1]
```diff
@@ -253,10 +253,26 @@ public static boolean attemptBindSingleValue(
...
+    boolean valueSet = false;
+
+    if (propertyType == UUID.class) {
+      if (propertyValue != null) {
+        com.google.protobuf.Value protoValue = com.google.protobuf.Value.newBuilder()
+            .setStringValue(propertyValue.toString())
+            .build();
+        valueBinder.to(Value.untyped(protoValue));
+      } else {
+        valueBinder.to((String) null);
```
Comment (olavloite) [inline spring-cloud-gcp-data-spanner/src/main/java/com/google/cloud/spring/data/spanner/core/convert/ConverterAwareMappingSpannerEntityWriter.java]: This will bind it to a __typed__ string. This should use the same trick as above; create a protovalue containing a null value, and then bind it as an untyped value.

We should also have a test that verifies that.

File: spring-cloud-gcp-data-spanner/src/test/java/com/google/cloud/spring/data/spanner/core/convert/ConverterAwareMappingSpannerEntityWriterTests.java [hunk 2]
```diff
@@ -319,7 +319,10 @@ void writeTest() {
     verify(bigDecimalsBinder, times(1)).toNumericArray(t.bigDecimals);
     verify(intervalFieldBinder, times(1)).to(t.intervalField);
     verify(intervalListFieldBinder, times(1)).toIntervalArray(t.intervalList);
-    verify(uuidFieldBinder, times(1)).to(t.uuidField);
+    com.google.protobuf.Value expectedProtoValue = com.google.protobuf.Value.newBuilder()
+        .setStringValue(t.uuidField.toString())
+        .build();
+    verify(uuidFieldBinder, times(1)).to(Value.untyped(expectedProtoValue));
```
Comment (olavloite) [inline spring-cloud-gcp-data-spanner/src/test/java/com/google/cloud/spring/data/spanner/core/convert/ConverterAwareMappingSpannerEntityWriterTests.java:325]: Add a test here that it does the same for NULL values.

File: spring-cloud-gcp-data-spanner/src/test/java/com/google/cloud/spring/data/spanner/repository/it/SpannerUuidIntegrationTests.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
+package com.google.cloud.spring.data.spanner.repository.it;
```
Comment (olavloite) [inline spring-cloud-gcp-data-spanner/src/test/java/com/google/cloud/spring/data/spanner/repository/it/SpannerUuidIntegrationTests.java:17]: nit: copyright header

File: spring-cloud-gcp-data-spanner/src/test/java/com/google/cloud/spring/data/spanner/repository/it/SpannerUuidIntegrationTests.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.condition.EnabledIfSystemProperty;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.test.context.junit.jupiter.SpringExtension;
+
+@EnabledIfSystemProperty(named = "it.spanner", matches = "true")
+@ExtendWith(SpringExtension.class)
+public class SpannerUuidIntegrationTests extends AbstractSpannerIntegrationTest {
```
Comment (olavloite) [inline spring-cloud-gcp-data-spanner/src/test/java/com/google/cloud/spring/data/spanner/repository/it/SpannerUuidIntegrationTests.java:38]: Also here: Add tests for null values

File: spring-cloud-gcp-data-spanner/src/test/java/com/google/cloud/spring/data/spanner/repository/it/SpannerUuidIntegrationTests.java (hunk 3 above)
Comment (olavloite) [reply spring-cloud-gcp-data-spanner/src/test/java/com/google/cloud/spring/data/spanner/repository/it/SpannerUuidIntegrationTests.java:17]: This also applies to the other new files that are added in this pull request

File: spring-cloud-gcp-data-spanner/src/test/java/com/google/cloud/spring/data/spanner/core/convert/ConverterAwareMappingSpannerEntityWriterTests.java (hunk 2 above)
Comment (sakthivelmanii) [reply spring-cloud-gcp-data-spanner/src/test/java/com/google/cloud/spring/data/spanner/core/convert/ConverterAwareMappingSpannerEntityWriterTests.java:325]: Done

File: spring-cloud-gcp-data-spanner/src/test/java/com/google/cloud/spring/data/spanner/repository/it/SpannerUuidIntegrationTests.java (hunk 3 above)
Comment (sakthivelmanii) [reply spring-cloud-gcp-data-spanner/src/test/java/com/google/cloud/spring/data/spanner/repository/it/SpannerUuidIntegrationTests.java:17]: Done

File: spring-cloud-gcp-data-spanner/src/test/java/com/google/cloud/spring/data/spanner/repository/it/SpannerUuidIntegrationTests.java (hunk 4 above)
Comment (sakthivelmanii) [reply spring-cloud-gcp-data-spanner/src/test/java/com/google/cloud/spring/data/spanner/repository/it/SpannerUuidIntegrationTests.java:38]: Done

File: spring-cloud-gcp-data-spanner/src/main/java/com/google/cloud/spring/data/spanner/core/convert/ConverterAwareMappingSpannerEntityWriter.java (hunk 1 above)
Comment (sakthivelmanii) [reply spring-cloud-gcp-data-spanner/src/main/java/com/google/cloud/spring/data/spanner/core/convert/ConverterAwareMappingSpannerEntityWriter.java]: Done

**Review by olavloite — APPROVED** [review]

**Review by jinseopkim0 — APPROVED** [review]
